scripts/run_unit_tests.py

```diff
@@ -0,0 +1,157 @@
+    return cmd_output.strip()
+
+
+def get_package_directories():
+    """Get a list of directories containing sub-packages.
+
+    :rtype: list
+    :returns: A list of all sub-package directories.
+    """
+    # Run ls-tree with
+    #          -d: For directories only
+    # --name-only: No extra info like the type/hash/permission bits.
+    # --full-name: Give path relative to root, rather than cwd.
+    ls_tree_out = check_output(
+        'git', 'ls-tree',
+        '-d', '--name-only', '--full-name',
+        'HEAD', PROJECT_ROOT)
+    result = []
+    for package in ls_tree_out.split('\n'):
+        if package not in IGNORED_DIRECTORIES:
+            result.append(package)
+    return result
+
+
+def run_package(package, tox_env):
+    """Run tox environment for a given package.
+
+    :type package: str
+    :param package: The name of the subdirectory which holds the sub-package.
+                    This will be a path relative to ``PROJECT_ROOT``.
+
+    :type tox_env: str
+    :param tox_env: The ``tox`` environment(s) to run in each sub-package.
+
+    :rtype: bool
+    :returns: Flag indicating if the test run succeeded.
+    """
+    curr_dir = os.getcwd()
+    package_dir = os.path.join(PROJECT_ROOT, package)
+    try:
+        os.chdir(package_dir)
+        return_code = subprocess.call(['tox', '-e', tox_env])
+        return return_code == 0
+    finally:
+        os.chdir(curr_dir)
+
+
+def get_parser():
+    """Get simple ``argparse`` parser to determine configuration.
+
+    :rtype: :class:`argparse.ArgumentParser`
+    :returns: The parser for this script.
+    """
+    description = 'Run tox environment(s) in all sub-packages.'
+    parser = argparse.ArgumentParser(description=description)
+    parser.add_argument(
+        '--tox-env-dir', dest='tox_env_dir',
+        help='The current tox environment directory.')
+    return parser
+
+
+def get_tox_env():
+    """Get the environment to be used with ``tox``.
+
+    :rtype: str
+    :returns: The current ``tox`` environment to be used, e.g. ``"py27"``.
+    """
+    parser = get_parser()
+    args = parser.parse_args()
+    env_dir = args.tox_env_dir
+    _, tox_env = os.path.split(env_dir)
+    tox_env = ENV_REMAP.get(tox_env, tox_env)
+    return tox_env
+
+
+def main():
+    """Run all the unit tests that need to be run."""
+    packages_to_run = get_package_directories()
+    if not packages_to_run:
+        print('No tests to run.')
+        return
+
+    tox_env = get_tox_env()
+    failed_packages = []
+    for package in packages_to_run:
+        succeeded = run_package(package, tox_env)
+        if not succeeded:
+            failed_packages.append(package)
+
+    if failed_packages:
+        msg_parts = ['Sub-packages failed:']
+        for package in failed_packages:
+            msg_parts.append('- ' + package)
+        msg = '\n'.join(msg_parts)
+        print(msg, file=sys.stderr)
+        sys.exit(len(failed_packages))
```

Comment:
Actually this is probably a better list...
http://www.virtsync.com/c-error-codes-include-errno